pyiceberg/io/pyarrow.py

```diff
@@ -698,77 +708,147 @@ def before_field(self, field: pa.Field) -> None:
+        return all(field_results)
+
+    def field(self, field: pa.Field, field_result: bool) -> bool:
+        return all([_get_field_id(field) is not None, field_result])
+
+    def list(self, list_type: pa.ListType, element_result: bool) -> bool:
         element_field = list_type.value_field
         element_id = _get_field_id(element_field)
-        if element_result is not None and element_id is not None:
-            return ListType(element_id, element_result, element_required=not element_field.nullable)
-        return None
+        return element_result and element_id is not None
 
-    def map(
-        self, map_type: pa.MapType, key_result: Optional[IcebergType], value_result: Optional[IcebergType]
-    ) -> Optional[IcebergType]:
+    def map(self, map_type: pa.MapType, key_result: bool, value_result: bool) -> bool:
         key_field = map_type.key_field
         key_id = _get_field_id(key_field)
         value_field = map_type.item_field
         value_id = _get_field_id(value_field)
-        if key_result is not None and value_result is not None and key_id is not None and value_id is not None:
-            return MapType(key_id, key_result, value_id, value_result, value_required=not value_field.nullable)
-        return None
+        return all([key_id is not None, value_id is not None, key_result, value_result])
+
+    def primitive(self, primitive: pa.DataType) -> bool:
+        return True
 
-    def primitive(self, primitive: pa.DataType) -> IcebergType:
+
+class _ConvertToIceberg(PyArrowSchemaVisitor[Union[IcebergType, Schema]]):
+    """Converts PyArrowSchema to Iceberg Schema. Applies the IDs from name_mapping if provided."""
+
+    field_names: List[str]
+
+    def __init__(self, name_mapping: Optional[NameMapping] = None) -> None:
+        self.field_names = []
+        self.name_mapping = name_mapping
+
+    def _current_path(self) -> str:
+        return ".".join(self.field_names)
+
+    def _path(self, name: str) -> str:
+        return ".".join(self.field_names + [name])
+
+    def _get_field_id(self, field: pa.Field) -> int:
+        if self.name_mapping:
+            return self.name_mapping.find(self._current_path()).field_id
+        elif field_id := _get_field_id(field):
+            return field_id
+        else:
+            raise ValueError(f"Cannot convert {field} to Iceberg Field as field_id is empty.")
+
+    def schema(self, schema: pa.Schema, struct_result: StructType) -> Schema:
+        return Schema(*struct_result.fields)
+
+    def struct(self, struct: pa.StructType, field_results: List[NestedField]) -> StructType:
+        return StructType(*field_results)
+
+    def field(self, field: pa.Field, field_result: IcebergType) -> NestedField:
+        field_id = self._get_field_id(field)
+        field_doc = _get_field_doc(field)
+        field_type = field_result
+        if field_type is None:
+            raise ValueError(f"Cannot convert {field} to Iceberg Field as field_type is empty.")
```

Comment:
Quick question: Could `field_type` ever be None, given that we raise `TypeError` for unsupported types? If so, should we consider marking `field_result` as `Optional[IcebergType]`?

Reply:
since it already raises a TypeError, I think we could make field_result IcebergType and just remove this check to make sure field_type has a value

Reply:
Thanks, I think removing this check is a good idea!